patches/server/0925-Paper-Plugins.patch

```diff
@@ -0,0 +1,1274 @@
++import java.lang.annotation.ElementType;
++import java.lang.annotation.Retention;
++import java.lang.annotation.RetentionPolicy;
++import java.lang.annotation.Target;
++import java.lang.reflect.Type;
++import java.util.regex.Pattern;
++
++public final class PluginConfigConstraints {
++
++    @Documented
++    @Retention(RetentionPolicy.RUNTIME)
++    @Target(ElementType.FIELD)
++    public @interface PluginName {
++
++        final class Factory implements Constraint.Factory<PluginName, String> {
++
++            private static final Pattern VALID_NAME = Pattern.compile("^[A-Za-z0-9 _.-]+$");
++
++            @Override
++            public Constraint<String> make(PluginName data, Type type) {
++                return value -> {
++                    if (value != null && !VALID_NAME.matcher(value).matches()) {
++                        throw new SerializationException("name '" + value + "' contains invalid characters.");
++                    }
++                };
++            }
++        }
++    }
++
++    @Documented
++    @Retention(RetentionPolicy.RUNTIME)
++    @Target(ElementType.FIELD)
++    public @interface PluginNameSpace {
++
++        final class Factory implements Constraint.Factory<PluginName, String> {
++
++            @Override
++            public Constraint<String> make(PluginName data, Type type) {
++                return value -> {
++                    if (value != null && value.startsWith("org.bukkit.")) {
++                        throw new SerializationException("main may not be within the org.bukkit namespace");
++                    }
++                };
++            }
++        }
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/provider/configuration/type/PaperCommandMap.java b/src/main/java/io/papermc/paper/plugin/provider/configuration/type/PaperCommandMap.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..314cfd14b0632e4b4f7f2ec1dbd977b0d1e331ff
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/configuration/type/PaperCommandMap.java
+@@ -0,0 +1,70 @@
++package io.papermc.paper.plugin.provider.configuration.type;
++
++import com.google.common.collect.ImmutableList;
++import com.google.common.collect.ImmutableMap;
++import org.spongepowered.configurate.serialize.ScalarSerializer;
++import org.spongepowered.configurate.serialize.SerializationException;
++
++import java.lang.reflect.Type;
++import java.util.Map;
++import java.util.function.Predicate;
++
++public class PaperCommandMap {
++
++    public static final Serializer SERIALIZER = new Serializer();
++
++    private final Map<String, Map<String, Object>> map;
++
++    public PaperCommandMap() {
++        this.map = Map.of();
++    }
++
++    public PaperCommandMap(Map<?, ?> data) throws SerializationException {
++        ImmutableMap.Builder<String, Map<String, Object>> commandsBuilder = ImmutableMap.builder();
++        try {
++            for (Map.Entry<?, ?> command : data.entrySet()) {
```

Comment:
This constructor is too complex delegate on another function and use "continue" to avoid large nesting stats here

Reply:
This is directly copied from PluginDescriptionFile#loadMap, cleaning this up might come later.